### DISCUSSION TITLE
Mermaid config issue

### DISCUSSION THREAD

**Discussion by skonik** [squidfunk/mkdocs-material#6984] [Q&A] [answered]
Hello. 

I would like to render my mermaid diagram with the config like on the picture below:
<img width="714" alt="Screenshot 2024-03-28 at 21 04 06" src="https://github.com/squidfunk/mkdocs-material/assets/50069473/cb7dce37-f00f-4510-814a-7d719d90d51b">

However mkdocs-materials ignores config section

<img width="464" alt="Screenshot 2024-03-28 at 21 05 37" src="https://github.com/squidfunk/mkdocs-material/assets/50069473/5e8c4954-fea4-4cb2-850b-e9ba9475183e">

<img width="335" alt="Screenshot 2024-03-28 at 21 05 24" src="https://github.com/squidfunk/mkdocs-material/assets/50069473/5bcc468f-4546-4c58-a9ab-6cb6453c7892">

Is it possible somehow to enable config parsing?

**Comment by alexvoss** [comment]
Try removing the `config:` element like so:

  ```
  ---
  flowchart:
    curve: stepAfter
  ---
  flowchart TD
   A[GE Solution] --- recEmpty[ ]:::empty
   B[GA Solution] --- recEmpty
   recEmpty --> C[Stir 10 mins at 500 r/min]
A> classDef empty width:0p
```

**Reply by skonik** [reply]
Still no effect. 

```
---
flowchart: 
  curve: stepAfter
---
flowchart TD
   A[GE Solution] --- recEmpty[ ]:::empty
   B[GA Solution] --- recEmpty
   recEmpty --> C[Stir 10 mins at 500 r/min]
   classDef empty width:0px
```

<img width="350" alt="Screenshot 2024-03-29 at 14 51 41" src="https://github.com/squidfunk/mkdocs-material/assets/50069473/374a5a94-387b-4d70-a67b-43ba55685803">

**Reply by alexvoss** [reply]
This is what I get:

<img width="263" alt="image" src="https://github.com/squidfunk/mkdocs-material/assets/4134224/4751b320-8b74-4b85-836a-00b0be581f96">

Is that what you are looking for? Sorry, I am not that familiar with Mermaid, having only dabbled with it.

**Reply by skonik** [reply]
I would like to make edges straight(it's circular now). 
So I found some examples on [stackoverflow](https://stackoverflow.com/questions/51367234/fusing-arrows-sideways-in-mermaid-diagrams) how to do it. 


This is what I would like to get:
```
---
config:
  flowchart: 
    curve: stepAfter
  
---
flowchart TD
   A[GE Solution] --- recEmpty[ ]:::empty
   B[GA Solution] --- recEmpty
   recEmpty --> C[Stir 10 mins at 500 r/min]
   classDef empty width:0px

```
<img width="497" alt="Screenshot 2024-03-29 at 19 19 23" src="https://github.com/squidfunk/mkdocs-material/assets/50069473/2fc76373-1d9b-4216-acb8-0763176f604c">

But using the same code inside mkdocs doesn't make any changes.  It's still circular.
I was wondering if someone knows what's the issue with the config. It works perfectly fine if I use [mermaid online editor](https://mermaid.live/edit#pako:eNp1j71uwzAMhF-F4Byj7tBFQwG3DvIA7mZ5ECQ6FmpJhkTlB0HevbKAduvG430E7x6ogyEU2DSN9Dr42Z6F9ADzGq56UZEF7BJA53ghAYlp62amuG-lr2d_LHz1Fe7G0xGGsGa2wU9QIIikj27j-wiTEIL2saIf46n7B63-ryjOO3yOA9sIry046xMohre2hfhS1FRpvaqUepqhPoCrNbyIdrvhAR1Fp6wpVR81OvJCjiSWfmhU_JYo_bNwKnMY7l6j4JjpgHkziqm36hyVQzGrNdHzB6l_Zjk)

Anyway thanks for your participation

**Reply by alexvoss** [reply]
Right, I was misled by the funny effect that the arrows do not meet up in your version.

**Reply by alexvoss** [reply]
I have pasted the output back into the Mermaid editor and it renders fine - after replacing the `&gt;` HTML entity with the `>`character. So, it does not seem to be the case that the Mermaid code gets mangled. I also turned on the debugger in the browser and checked [what was being passed into Mermaid for rendering](https://github.com/squidfunk/mkdocs-material/blob/f27b93ece3c423537873e0bc5de55b3c3381792f/src/templates/assets/javascripts/components/content/mermaid/index.ts#L116):

```
---\nconfig: \n  flowchart:\n    curve: stepAfter\n---\nflowchart TD\n A[GE Solution] --- recEmpty[ ]:::empty\n B[GA Solution] --- recEmpty\n recEmpty --> C[Stir 10 mins at 500 r/min]\n classDef empty width:0px
```

That looks fine to me. Also bumped up the Mermaid version locally to 10.9.0 and it made no difference. Does not seem to be a version issue. Maybe I got something wrong somewhere or did not manage to look in the right places.

**Reply by kamilkrzyskow** [reply, marked as answer]
Works for me @skonik with newest theme version 9.5.15 with mermaid 10.7. The frontmatter config is available since version 10.4 
Maybe you're using some older version of the material theme, which uses an older mermaid version?

[mermaid-config-frontmatter.zip](https://github.com/squidfunk/mkdocs-material/files/14807721/mermaid-config-frontmatter.zip)

![image](https://github.com/squidfunk/mkdocs-material/assets/34622465/589646ac-eacc-42f7-9969-4d119e8b8b2e)

EDIT:
Also thanks for letting me know about the frontmatter config, we opted out of mermaid because of those curvy lines which were unconfigurable, good to know there is a way to control it from Markdown now.

**Reply by alexvoss** [reply]
Bloody hell. Your version works for me as well. I have just removed what I had by reusing it for another issue and deleting the crucial bits. Argh. I am using 9.5.15, public version.

**Reply by kamilkrzyskow** [reply]
Looking at the changelog, before version https://squidfunk.github.io/mkdocs-material/changelog/#9.4.13 Material used mermaid 9.4, so there was no config frontmatter support.

This could truly be a groundbreaking discovery for me, as it might influence the need for shadow root opening:
- https://github.com/squidfunk/mkdocs-material/discussions/6778#discussioncomment-8751560

---

> also turned on the debugger in the browser and checked [what was being passed into Mermaid for rendering](https://github.com/squidfunk/mkdocs-material/blob/f27b93ece3c423537873e0bc5de55b3c3381792f/src/templates/assets/javascripts/components/content/mermaid/index.ts#L116):

Before running the example, I also checked the minified code with the debugger on the Live Mermaid site and it also uses the `mermaid.render(id, content)` function.
The theme's implementation uses the textContent and doesn't sanitize anything, so I just assuemed it will work and run the example from the comment :v:

**Reply by skonik** [reply]
Wow. Yeah, I've been using older version of `mkdocs-material`. Thank you!

**Reply by alexvoss** [reply]
So, I don't have any experience with the whole shadow DOM stuff but if I understand correctly, it would be possible to access the contents from JavaScript if it was open. The code would need to use `element.shadowRoot` explicitly to get access, while none of the `document.getElement` and `document.querySelector` methods would select content in the shadow DOM. Neither would CSS selectors. Previous discussions seem to suggest that `closed` was used out of an abundance of caution. So, is the question how we might ascertain whether opening it up would cause any issues? I don't see where they would be coming from but then I am neither a web developer nor a chap with much imagination ;o)

**Reply by kamilkrzyskow** [reply]
@alexvoss 
As I stated in https://github.com/squidfunk/mkdocs-material/discussions/6778#discussioncomment-8751560 on pymdownx mkdocs-material site, custom diagrams with open shadow roots are being used, so I believe we can simply switch. 
But as per usual with such things best to be extra cautious, and check all 4 variants, open/closed, loose/strict security, all within the Material for MkDocs website. 
Apart of the `document.querySelector` we'd need to check how ids behave with plain hreftags, aka if the mermaid has an ID that is the same as some TOC item, will it properly move to the main root, or will it jump to the already rendered mermaid.
The task isn't so complicated, some GPT AI could likely generate most of the test code, but it requires an organized testing procedure, and jotting down a checklist, I'm a bit short on time currently, so can't really dive deeper.
